libs/metrics-wai/src/Data/Metrics/Types.hs

```diff
@@ -0,0 +1,62 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE ViewPatterns        #-}
+
+module Data.Metrics.Types
+    ( PathTemplate(..)
+    , Paths(..)
+    , mkTree
+    , treeLookup
+    ) where
+
+import Imports
+import Data.Tree as Tree
+
+import qualified Data.ByteString.Char8 as BS
+
+
+newtype PathTemplate = PathTemplate Text
+
+-- | A 'Forest' of path segments.  A path segment is 'Nothing' if it captures a value
+-- (e.g. user id).
+newtype Paths = Paths (Forest (Maybe ByteString))
+  deriving (Eq, Show)
+
+-- | Turn a list of paths into a 'Paths' tree.  Treat all path segments that start with @':'@
+-- as equal and turn them into 'Nothing'.
+mkTree :: forall m. (m ~ Either String) => [[ByteString]] -> m Paths
+mkTree = fmap (Paths . melt) . mapM mkbranch . sortBy (flip compare) . fmap (fmap mknode)
+  where
+    mkbranch :: [Maybe ByteString] -> m (Tree (Maybe ByteString))
+    mkbranch (seg : segs@(_:_)) = Node seg . (:[]) <$> mkbranch segs
+    mkbranch (seg : [])         = Right $ Node seg []
+    mkbranch []                 = Left "internal error: path with on segments."
+
+    mknode :: ByteString -> Maybe ByteString
+    mknode seg = if BS.head seg /= ':' then Just seg else Nothing
+
+    melt :: Forest (Maybe ByteString) -> Forest (Maybe ByteString)
+    melt [] = []
+    melt (tree : []) = [tree]
+    melt (tree : tree' : trees) = if rootLabel tree == rootLabel tree'
+        then let tree'' = Node (rootLabel tree) (melt $ subForest tree <> subForest tree')
+             in melt (tree'' : trees)
+        else tree : melt (tree' : trees)
+
+-- | A variant of 'Network.Wai.Route.Tree.lookup'.  The segments contain values to be captured
+-- when running the 'App', but here we simply replace them with @"<>"@.
+treeLookup :: Paths -> [ByteString] -> Maybe ByteString
```

Comment:
Just for fun I built the mkTree/treeLookup thing with Cofree and a custom datatype; remind me tomorrow and I'll walk you through Cofree (just for fun 😄 )

It's probably a bit confusing to read (maybe best re-written with a simple recursive data-type rather than Cofree); but I believe it fixes all the bugs we discovered yesterday evening. That said, it looks like you took another pass at this one, so if yours is working, probably best to stick with it 👍  mine is unnecessarily complicated.

https://gist.github.com/ChrisPenner/829279ef928512b84ce1448241f90aa8

Reply:
I would say it's simpler, and it's definitely prettier!  :)

But demands more general Haskell knowledge to be understood, so I'm slightly in favor of sticking with it.  Happy to be overruled, though!